README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# `leaflet.circlesector`: An R wrapper for the Leaflet.Circle-sector plugin <img src="man/figures/logo.jpg" align="right" alt="A hatched nest with a fan of turtle hatchling tracks oritented towards the ocean" width="120" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/dbca-wa/leaflet.circlesector/workflows/R-CMD-check/badge.svg)](https://github.com/dbca-wa/leaflet.circlesector/actions)
<!-- badges: end -->

The R package `leaflet.circlesector` wraps the Leaflet plugin
[Leaflet.Circle-sector](https://github.com/kluizeberg/Leaflet.Circle-sector),
providing a way to draw circle segments on leaflet maps.

This can be used to visualise marine turtle hatchling (mis)orientation
at the time of nest emergence.

See
[`vignette("Introduction")`](https://dbca-wa.github.io/leaflet.circlesector/articles/intro.html)
for an explanation of packaged example data and desired visualisation.

``` r
library(leaflet.circlesector)
data("fans_tracks")
data("fans_mean")
data("fans_water")
data("outlier_segments")

leaflet_basemap(l_height = 500, l_width = 700) %>% 
    leaflet::clearBounds() %>% 
    leaflet::addCircles(
        data = fans_tracks,
        lat = ~ lat,
        lng = ~ lon,
        color = 'white',
        weight = 2,
        radius = 5
    ) %>%
    addCircleSectorMid(data = fans_mean) %>%
    addCircleSectorMid(data = fans_water) %>%
    addCircleSectorMid(data = outlier_segments) %>%
    addCircleSectorMinMax(data = fans_tracks) %>% 
    leaflet::setView(115.0245, -21.46, 19)
```

<img src="man/figures/README-example_map-1.png" width="100%" />

## Installation

You can install the development version of `leaflet.circlesector` from
[GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("dbca-wa/leaflet.circlesector")
```


NEWS.md
# leaflet.circlesector 0.1.0
* Set bounds to extent of data for `addCircleSectorMinMax` and `addCircleSectorMid`
* Update examples and vignette (call `leaflet::clearBounds` to release `setView` 
  of `leaflet_basemap`).


tests/testthat.R
library(testthat)
library(leaflet.circlesector)

test_check("leaflet.circlesector")


tests/testthat/test-addCircleSector.R
test_that("addCircleSectorMinMax returns a leaflet map", {
    data("fans_tracks")
    themap <- leaflet_basemap(l_height = 500, l_width = 700) %>%
        addCircleSectorMinMax(data = fans_tracks)

    expect_s3_class(themap, "leaflet")
})

test_that("addCircleSectorMid returns a leaflet map", {
    data("fans_mean")
    themap <- leaflet_basemap(l_height = 500, l_width = 700) %>%
        addCircleSectorMid(data = fans_mean)

    expect_s3_class(themap, "leaflet")
})


tests/testthat/test-leaflet_basemap.R
test_that("leaflet_basemap works", {
  themap <- leaflet_basemap()
  expect_s3_class(themap, "leaflet")
})


R/data.R
#' Turtle hatchling nest emergences: Main fan of tracks
#'
#' Nest fans are described by
#' * encounter_source_id A UUID for the TurtleNestEncounter
#' * encounter_when Datetime of the encounter
#' * encounter_latitude Latitude of the encounter in decimal degrees WGS84
#' * encounter_laongitude Longitude of the encounter in decimal degrees WGS84
#' * encounter_name The nest tag if known, a unique ID for the nest
#' * encounter_species The species of turtle which laid that nest
#' * bearing_leftmost_track_degrees The leftmost fan angle bearing
#' * bearing_rightmost_track_degrees The rightmost fan angle bearing
#' * no_tracks_main_group The number of hatchling tracks in the main group of
#'   tracks, best estimate
#' * no_tracks_main_group_min The number of hatchling tracks in the main group of
#'   tracks, lowest estimate
#' * no_tracks_main_group_max The number of hatchling tracks in the main group of
#'   tracks, highest estimate
#' * outlier_tracks_present Whether outlier tracks were present
#' * path_to_sea_comments Any comments on the path to sea, e.g. whether obstructed
#' * hatchling_emergence_time_known Whether the exact time of the emergence is known
#' * light_sources_present
#' * hatchling_emergence_time
#' * hatchling_emergence_time_accuracy
#' * cloud_cover_at_emergence
#' * datetime
#' * calendar_date_awst
#' * calendar_date_awst_text
#' * calendar_year
#' * turtle_date
#' * turtle_date_awst_text
#' * season
#' * season_week
#' * iso_week
#'
#' @family included
"fans"

#' Turtle hatchling nest emergences: Individual outlier tracks
#'
#' With the same base columns as in dataset fans:
#' * bearing_outlier_track_degrees The compass bearing of the outlier track 5m
#'   from the nest in degrees
#' * outlier_group_size The number of tracks in the outlier group, typically 1
#' * outlier_track_comment Any further comments on the outlier track
#'
#' @family included
"outliers"

#' Turtle hatchling nest emergences: Known light sources
#'
#' With the same base columns as fans:
#' * bearing_light_degrees
#' * light_source_type
#' * light_source_description
#'
#' @family included
"lights"


#' A data frame ready to be used in `addCircleSectorMinMax`
#'
#' This data frame is created from `fans` in `inst/generate_data.R`.
#'
#' Columns:
#' * lat
#' * lon
#' * start_angle
#' * end_angle
#' * radius
#' * weight
#' * colour
#' * popup
#'
#' @family included
"fans_tracks"


#' A data frame ready to be used in `addCircleSectorMid`
#'
#' #' This data frame is created from `fans` in `inst/generate_data.R`.
#'
#' Columns:
#' * lat
#' * lon
#' * mid
#' * angle
#' * radius
#' * weight
#' * colour
#' * popup
#'
#' @family included
"fans_mean"



#' A data frame ready to be used in `addCircleSectorMid`
#'
#' #' This data frame is created from `fans` in `inst/generate_data.R`.
#'
#' Columns:
#' * lat
#' * lon
#' * mid
#' * angle
#' * radius
#' * weight
#' * colour
#' * popup
#'
#' @family included
"fans_water"


#' A data frame ready to be used in `addCircleSectorMid`
#'
#' This data frame is created from `fans` in `inst/generate_data.R`.
#'
#' Columns:
#' * lat
#' * lon
#' * mid
#' * angle
#' * radius
#' * weight
#' * colour
#' * popup
#'
#' @family included
"outlier_segments"


#' A data frame ready to be used in `addCircleSectorMid`
#'
#' This data frame is created from `fans` in `inst/generate_data.R`.
#'
#' Columns:
#' * lat
#' * lon
#' * mid
#' * angle
#' * radius
#' * weight
#' * colour
#' * popup
#'
#' @family included
"light_segments_artificial"


#' A data frame ready to be used in `addCircleSectorMid`
#'
#' This data frame is created from `fans` in `inst/generate_data.R`.
#'
#' Columns:
#' * lat
#' * lon
#' * mid
#' * angle
#' * radius
#' * weight
#' * colour
#' * popup
#'
#' @family included
"light_segments_natural"


R/leaflet_basemap.R
#' Create a Leaflet basemap focused on Western Australia
#'
#' @param l_width The leaflet map width, default: NULL
#' @param l_height The leaflet map height, default: NULL
#'
#' @return A Leaflet map focused on WA
#' @export
#' @family plugin
#'
#' @examples
#' leaflet_basemap()
leaflet_basemap <- function(l_width = NULL, l_height = NULL) {
    leaflet::leaflet(width = l_width, height = l_height) %>%
        leaflet::addProviderTiles("Esri.WorldImagery", group = "Basemap") %>%
        leaflet::addProviderTiles(
            "OpenStreetMap.Mapnik",
            group = "Basemap",
            options = leaflet::providerTileOptions(opacity = 0.35)
        ) %>%
        leaflet.extras::addFullscreenControl(pseudoFullscreen = TRUE) %>%
        leaflet::addScaleBar(
            position = "bottomleft",
            options = leaflet::scaleBarOptions(
                imperial = FALSE,
                maxWidth = 200
            )
        ) %>%
        leaflet::addMiniMap(toggleDisplay = TRUE) %>%
        leaflet::setView(130, -20, 5)
}

# usethis::use_test("leaflet_basemap")  # nolint


inst/generate_data.R
library(wastdr)
data("wastd_data")
data_thv <- wastd_data %>%
  wastdr::filter_wastd_turtledata(area_name = "Thevenard Island")

fans <- data_thv$nest_fans %>% wastdr::filter_realspecies()
outliers <- data_thv$nest_fan_outliers %>% wastdr::filter_realspecies()
lights <- data_thv$nest_lightsources %>% wastdr::filter_realspecies()

usethis::use_data(fans, compress = "xz", overwrite = TRUE)
usethis::use_data(outliers, compress = "xz", overwrite = TRUE)
usethis::use_data(lights, compress = "xz", overwrite = TRUE)


# Transmute data to default column names and values (radius, colour)
fans_tracks <- fans %>%
  dplyr::transmute(
    lat = encounter_latitude,
    lon = encounter_longitude,
    start_angle = bearing_leftmost_track_degrees,
    end_angle = bearing_rightmost_track_degrees,
    radius = 10,
    weight = 2,
    colour = "blue",
    label = glue::glue(
      "{no_tracks_main_group} tracks, ",
      "path: {stringr::str_replace_all(path_to_sea_comments, 'None', '')}"
    ),
    popup = glue::glue(
      "<h3>{no_tracks_main_group} tracks ",
      "<small>{no_tracks_main_group_min}-{no_tracks_main_group_max}</small></h3>",
      "Path to sea: {stringr::str_replace_all(path_to_sea_comments, 'None', '')}"
    )
  )

# The main direction to water is a 20m blue line
fans_mean <- fans %>%
  dplyr::transmute(
    lat = encounter_latitude,
    lon = encounter_longitude,
    bearing = bearing_leftmost_track_degrees + (
      (bearing_leftmost_track_degrees - bearing_rightmost_track_degrees) / 2),
    angle = 1,
    radius = 15,
    weight = 1,
    colour = "blue",
    label = glue::glue("Mean bearing: {bearing} deg"),
    popup = glue::glue("Mean bearing: {bearing} deg")
  )


# The main direction to water is a 20m black line
fans_water <- fans %>%
  dplyr::transmute(
    lat = encounter_latitude,
    lon = encounter_longitude,
    bearing = bearing_to_water_degrees,
    angle = 1,
    radius = 15,
    weight = 1,
    colour = "black",
    label = glue::glue("Bearing to water: {bearing} deg"),
    popup = glue::glue("Bearing to water: {bearing} deg")
  )


# Outlier tracks are 12m red lines
outlier_segments <- outliers %>%
  dplyr::transmute(
    lat = encounter_latitude,
    lon = encounter_longitude,
    bearing = bearing_outlier_track_degrees,
    angle = 2,
    radius = 12,
    weight = 1,
    colour = "red",
    label = glue::glue("{outlier_group_size} track(s) {bearing} deg {outlier_track_comment}"),
    popup = glue::glue("<h3>{outlier_group_size} track(s)</h3>",
                       "Bearing: {bearing} deg<br/>",
                       "{outlier_track_comment}")
  )


# Known light sources are 100m orange lines
light_segments_artificial <- lights %>%
  dplyr::filter(light_source_type == "artificial") %>%
  dplyr::transmute(
    lat = encounter_latitude,
    lon = encounter_longitude,
    bearing = bearing_light_degrees,
    angle = 1,
    radius = 100,
    weight = 2,
    colour = "#FFC300",
    label = glue::glue("{light_source_description}"),
    popup = glue::glue("<h3>{light_source_description}</h3>",
                       "Artifical light source<br/>",
                       "Bearing {bearing} deg")
  )

light_segments_natural <- lights %>%
  dplyr::filter(light_source_type == "natural") %>%
  dplyr::transmute(
    lat = encounter_latitude,
    lon = encounter_longitude,
    bearing = bearing_light_degrees,
    angle = 1,
    radius = 100,
    weight = 2,
    colour = "#f0ebc2",
    label = glue::glue("{light_source_description}"),
    popup = glue::glue("<h3>{light_source_description}</h3>",
                       "Natural light source<br/>",
                       "Bearing {bearing} deg")
  )

usethis::use_data(fans_tracks, compress = "xz", overwrite = TRUE)
usethis::use_data(fans_mean, compress = "xz", overwrite = TRUE)
usethis::use_data(fans_water, compress = "xz", overwrite = TRUE)
usethis::use_data(outlier_segments, compress = "xz", overwrite = TRUE)
usethis::use_data(light_segments_artificial, compress = "xz", overwrite = TRUE)
usethis::use_data(light_segments_natural, compress = "xz", overwrite = TRUE)

curl::curl_download(
  url = "https://raw.githubusercontent.com/kluizeberg/Leaflet.Circle-sector/master/leaflet.circle-sector.js",
  destfile = here::here("inst/Leaflet.Circle-sector-1.6/leaflet.circle-sector.js")
)


inst/generate_hexlogo.R
# -----------------------------------------------------------------------------#
# Hex sticker
#
# remotes::install_github("GuangchuangYu/hexSticker")
# library(hexSticker)
# library(showtext)
sysfonts::font_add_google("Knewave", "knewave")
sysfonts::font_add_google("Prosto One", "prosto")
# sysfonts::font_add_google("Concert One", "concert")
turtle <- here::here("man", "figures", "fanangles.jpg")
logo <- here::here("man", "figures", "logo.jpg")
darkred <- "#a50b0b"
# logo s_, text p_, bg h_
hexSticker::sticker(
    turtle,
    # asp = 0.684, dpi = 300,
    s_x = 1.0, s_y = 1.00, s_width = 1.3, # s_height = 0.1,
    package = "leaflet.circlesector", p_x = 1, p_y = 1.2, p_size = 12,
    p_family = "prosto", p_color = darkred,
    h_fill = "#aaaaaa", h_color = darkred,
    # url = "docs.ropensci.org/ruODK", u_size = 4, u_color = "#ffffff",
    white_around_sticker = T,
    filename = logo
)


R/circleSector.R
#' Add Circle sectors to a Leaflet map from a dataframe using start and end
#' angles
#'
#' This methor wraps the `setAngles` method of
#' [Leaflet.Circle-sector](https://github.com/kluizeberg/Leaflet.Circle-sector),
#' a Leaflet plugin extending
#' [L.Circle](https://leafletjs.com/index.html#circle).
#'
#'
#' @param map A leaflet map
#' @param data A data frame where each row represents on circle sector and
#'   columns for lat, lon, min, mid, max.
#' @param lat The column name as string holding the Latitude of the circle mid
#'   point on the map in WGS 84 decimal degrees. Default: "lat".
#' @param lon The column name as string holding the Longitude of the circle mid
#'   point on the map in WGS 84 decimal degrees. Default: "lon".
#' @param start_angle The column name as string holding the startAngle of the sector
#'   in decimal degrees, counted clockwise from North. Default: "start_angle".
#' @param end_angle The column name as string holding the endAngle of the sector
#'   in decimal degrees, counted clockwise from North. Default: "end_angle".
#' @param radius The column name as string holding the radius of the circle in
#'   metres. Default: "radius".
#' @param weight The column name as string holding the weight of the circle.
#'   Default: "weight.
#' @param colour The column name as string holding the colour of the circle.
#'   Default: "colour". Example values: "blue", "red", "#3388ff".
#' @param label The column name as string holding the label content of the circle.
#'   Default: "label".
#' @param popup The column name as string holding the popup content of the circle.
#'   Default: "popup".
#' @export
#' @family plugin
#' @examples
#' \dontrun{
#' data("fans_tracks")
#'
#' leaflet_basemap(l_height = 500, l_width = 700) %>%
#'   leaflet::clearBounds() %>%
#'   addCircleSectorMinMax(data = fans_tracks)
#' }
addCircleSectorMinMax <- function(map,
                                  data,
                                  lat = "lat",
                                  lon = "lon",
                                  start_angle = "start_angle",
                                  end_angle = "end_angle",
                                  radius = "radius",
                                  weight = "weight",
                                  colour = "colour",
                                  label = "label",
                                  popup = "popup") {

  # TODO abort if required cols are missing in data
  # TODO warn and swap if max < min bearings
  # TODO warn and treat if (max - min) > 360


  # Interpolate column names in JS onRender function
  js_string <- glue::glue("
    function(el, x, data) {{
        for (var i = 0; i < data.{lon}.length; i++) {{
            L.circle(
                [data['{lat}'][i], data['{lon}'][i]], {{
	            color: data['{colour}'][i],
	            radius: data['{radius}'][i],
	            weight: data['{weight}'][i]
                }})
        .setAngles(data['{start_angle}'][i], data['{end_angle}'][i])
        .bindTooltip(data['{label}'][i])
    	.bindPopup(data['{popup}'][i])
    	.addTo(this);
    }}
 }}")

  map %>%
    registerPlugin(circlePlugin()) %>%
    htmlwidgets::onRender(js_string, data = data) %>%
    leaflet::expandLimits(dplyr::pull(data, lat), dplyr::pull(data, lon))
}

#' Add Circle sectors to a Leaflet map from a dataframe using bearing and spread
#'
#' This methor wraps the `setSector` method of
#' [Leaflet.Circle-sector](https://github.com/kluizeberg/Leaflet.Circle-sector),
#' a Leaflet plugin extending [L.Circle](https://leafletjs.com/index.html#circle).
#'
#' @param map A leaflet map
#' @param data A data frame where each row represents on circle sector and
#'   columns for lat, lon, min, mid, max.
#' @param lat The column name as string holding the Latitude of the circle mid
#'   point on the map in WGS 84 decimal degrees. Default: "lat".
#' @param lon The column name as string holding the Longitude of the circle mid
#'   point on the map in WGS 84 decimal degrees. Default: "lon".
#' @param bearing The column name as string holding the main bearing of the sector
#'   in decimal degrees, counted clockwise from North. Default: "bearing".
#' @param angle max The column name as string holding the angle of the sector
#'   in decimal degrees, which will be centered on the main direction `mid`.
#'   Default: "angle".
#' @param radius The column name as string holding the radius of the circle in
#'   metres. Default: "radius".
#' @param weight The column name as string holding the weight of the circle.
#'   Default: "weight.
#' @param colour The column name as string holding the colour of the circle.
#'   Default: "colour". Example values: "blue", "red", "#3388ff".
#' @param label The column name as string holding the label content of the circle.
#'   Default: "label".
#' @param popup The column name as string holding the popup content of the circle.
#'   Default: "popup".
#' @export
#' @export
#' @family plugin
#' @examples
#' \dontrun{
#' data("fans_mean")
#'
#' leaflet_basemap(l_height = 500, l_width = 700) %>%
#'   leaflet::clearBounds() %>%
#'   addCircleSectorMid(data = fans_mean)
#' }
addCircleSectorMid <- function(map,
                               data,
                               lat = "lat",
                               lon = "lon",
                               bearing = "bearing",
                               angle = "angle",
                               radius = "radius",
                               weight = "weight",
                               colour = "colour",
                               label = "label",
                               popup = "popup") {

  # TODO abort if required cols are missing in data
  # TODO warn and swap if max < min bearings
  # TODO warn and treat if mid outside [0:360]

  # Interpolate column names in JS onRender function
  js_string <- glue::glue("
    function(el, x, data) {{
        for (var i = 0; i < data.{lon}.length; i++) {{
            L.circle(
                [data['{lat}'][i], data['{lon}'][i]], {{
	            color: data['{colour}'][i],
	            radius: data['{radius}'][i],
	            weight: data['{weight}'][i]
                }})
        .setSector(data['{bearing}'][i], data['{angle}'][i])
        .bindTooltip(data['{label}'][i])
    	.bindPopup(data['{popup}'][i])
    	.addTo(this);
    }}
 }}")


  map %>%
    registerPlugin(circlePlugin()) %>%
    htmlwidgets::onRender(js_string, data = data) %>%
    leaflet::expandLimits(dplyr::pull(data, lat), dplyr::pull(data, lon))
}

# usethis::use_test("addCircleSector")  # nolint
